lang/src/org/partiql/lang/eval/visitors/AggregationVisitorTransform.kt

```diff
@@ -0,0 +1,364 @@
+        val transformedKeys = mutableListOf<PartiqlAst.GroupKey>()
+        val groupKeyInformation = node.group.keyList.keys.mapIndexed { index, key ->
+            val publicAlias = key.asAlias?.text ?: key.expr.extractColumnAlias(index)
+            val uniqueAlias = uniqueAlias(this.contextStack.size, index)
+            val represents = key.expr
+            val transformedKey = PartiqlAst.build { groupKey(transformExpr(key.expr), uniqueAlias, key.metas) }
+            transformedKeys.add(transformedKey)
+            val isPublicAliasUserDefined = key.asAlias != null
+            GroupKeyInformation(groupKey = key, publicAlias = publicAlias, uniqueAlias = uniqueAlias, represents = represents, isPublicAliasUserDefined = isPublicAliasUserDefined)
+        }
+
+        // Add to Context Stack and return modified Group Keys
+        val hasAggregateOperator = containsAggregations || groupKeyInformation.isNotEmpty() || groupAsAlias != null
+        val ctx = VisitorContext(groupKeyInformation, groupAsAlias, hasAggregateOperator)
+        contextStack.add(ctx)
+        return PartiqlAst.build {
+            groupBy(
+                strategy = transformGroupBy_strategy(node.group),
+                keyList = groupKeyList(transformedKeys),
+                groupAsAlias = groupAsAlias,
+                metas = node.group.metas
+            )
+        }
+    }
+
+    /**
+     * Replaces [node] with [PartiqlAst.Projection.ProjectList] IF there are Group Keys and/or a Group Alias
+     */
+    override fun transformProjectionProjectStar(node: PartiqlAst.Projection.ProjectStar): PartiqlAst.Projection {
+        if (contextStack.last().groupKeys.isNotEmpty() || contextStack.last().groupAsAlias != null) {
+            return PartiqlAst.build {
+                val projectionItems = contextStack.last().groupKeys.map { key ->
+                    projectExpr(id(key.uniqueAlias, caseSensitive(), unqualified()), key.publicAlias)
+                }.toMutableList()
+
+                contextStack.last().groupAsAlias?.let { alias ->
+                    val item = projectExpr(id(alias, caseSensitive(), unqualified()), alias)
+                    projectionItems.add(item)
+                }
+                projectList(projectionItems)
+            }
+        }
+        return super.transformProjectionProjectStar(node)
+    }
+
+    override fun transformProjectionProjectValue(node: PartiqlAst.Projection.ProjectValue): PartiqlAst.Projection =
+        PartiqlAst.build {
+            projectValue(
+                value = itemTransform.transformExpr(node.value),
+                metas = node.metas
+            )
+        }
+
+    override fun transformSortSpec_expr(node: PartiqlAst.SortSpec) = itemTransform.transformSortSpec_expr(node)
+
+    override fun transformExprSelect_having(node: PartiqlAst.Expr.Select): PartiqlAst.Expr? = node.having?.let { having ->
+        itemTransform.transformExpr(having)
+    }
```

Comment:
Nit: If possible, would you please change the order of the override functions to the actually evaluation order? i.e. move those two below the `transformProjectionProjectList`.